central/notifications/service/service_impl.go

```diff
@@ -0,0 +1,142 @@
+	count, err := s.ds.CountNotifications(ctx, query)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to count notifications")
+	}
+	return &v1.CountNotificationsResponse{Count: int64(count)}, nil
+}
+
+// GetNotification returns a specific notification based on its ID.
+func (s *serviceImpl) GetNotification(ctx context.Context, resource *v1.ResourceByID) (*v1.GetNotificationResponse, error) {
+	resourceID := resource.GetId()
+	notification, err := s.ds.GetNotificationByID(ctx, resourceID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get notification %q", resourceID)
+	}
+	return &v1.GetNotificationResponse{Notification: convertToServiceType(notification)}, err
+}
+
+// ListNotifications returns all notifications matching the request query.
+func (s *serviceImpl) ListNotifications(ctx context.Context, request *v1.ListNotificationsRequest) (*v1.ListNotificationsResponse, error) {
+	query := getQueryBuilderFromFilter(request.GetFilter()).ProtoQuery()
+	paginated.FillPagination(query, request.GetPagination(), maxPaginationLimit)
+	paginated.FillDefaultSortOption(
+		query,
+		&v1.QuerySortOption{
+			Field:    search.LastUpdatedTime.String(),
+			Reversed: true,
+		},
+	)
+
+	notifications, err := s.ds.ListNotifications(ctx, query)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list notifications")
+	}
+	var respNotifications []*v1.Notification
+	for _, n := range notifications {
+		respNotifications = append(respNotifications, convertToServiceType(n))
+	}
+	return &v1.ListNotificationsResponse{Notifications: respNotifications}, nil
+}
+
+func getQueryBuilderFromFilter(filter *v1.NotificationsFilter) *search.QueryBuilder {
+	queryBuilder := search.NewQueryBuilder().
+		AddTimeRangeField(
+			search.CreatedTime,
+			protoconv.ConvertTimestampToTimeOrDefault(filter.GetFrom(), time.Unix(0, 0)),
+			protoconv.ConvertTimestampToTimeOrNow(filter.GetUntil()),
+		)
+	if domain := filter.GetDomain(); domain != "" {
+		queryBuilder = queryBuilder.AddExactMatches(search.NotificationDomain, domain)
+	}
+	if level := filter.GetLevel(); level != v1.NotificationLevel_NOTIFICATION_LEVEL_UNKNOWN {
+		queryBuilder = queryBuilder.AddExactMatches(search.NotificationLevel, level.String())
+	}
+	if notificationType := filter.GetNotificationType(); notificationType != v1.NotificationType_NOTIFICATION_TYPE_UNKNOWN {
+		queryBuilder = queryBuilder.AddExactMatches(search.NotificationType, notificationType.String())
+	}
+	if resourceType := filter.GetResourceType(); resourceType != "" {
+		queryBuilder = queryBuilder.AddExactMatches(search.ResourceType, resourceType)
+	}
+	return queryBuilder
+}
+
+func convertToServiceType(notification *storage.Notification) *v1.Notification {
+	if notification == nil {
+		return nil
+	}
+	return &v1.Notification{
+		Id:           notification.GetId(),
+		Type:         v1.NotificationType(notification.GetType()),
+		Level:        v1.NotificationLevel(notification.GetLevel()),
```

Comment:
I'm afraid this is not enough. IIRC, because the underlying type is int32, so if storage type has more values than the service type, the v1 type receives an unknown value and service response will contain an empty string because of:
https://github.com/stackrox/stackrox/blob/29877d61b082894fc9a1269a78ca870d17023034/generated/api/v1/notification_service.pb.go#L68-L74

I suggest to explicitly map all unknown map values to `Unknown`. With v2 APIs on the horizon, maybe, we need a global helper somewhere that does this automagically.